### PR TITLE
[WIP] feat(test): refreshed test report

### PR DESCRIPTION
**Pull request by bartlomieju**
Screenshot 2022-04-16 at 22 03 38

Closes https://github.com/denoland/deno/issues/14190
Closes https://github.com/denoland/deno/issues/13913

### CODE REVIEW

File: cli/fmt_errors.rs [hunk 1]
```diff
@@ -220,25 +228,59 @@ fn format_js_error(js_error: &JsError, is_child: bool) -> String {
...
+          false
+        }
+      })
+      .collect();
+    js_error.frames = frames;
+
+    let pretty_js_error = Self {
+      js_error,
+      color_message: true,
+    };
```
Comment (bartlomieju) [inline cli/fmt_errors.rs:265]: @nayeemrmn does this seem fine?

Comment (nayeemrmn) [reply cli/fmt_errors.rs:265]: IMO this change shouldn't involve any change to `fmt_errors.rs`, we should have a function in `cli/tools/test.rs`:
```rust
fn format_test_error(js_error: &JsError) -> String {
  let mut js_error = js_error.clone();
  let e = colors::red(&js_error.exception_message).to_string();
  js_error.exception_message = e;
  let frames = std::mem::take(&mut js_error.frames);
  let frames = frames
    .into_iter()
    .rev()
    .skip_while(|f| {
      if let Some(file_name) = &f.file_name {
        file_name.starts_with("[deno:") || file_name.starts_with("deno:")
      } else {
        false
      }
    })
    .collect();
  js_error.frames = frames;
  PrettyJsError::create(js_error)
    .to_string()
    .trim_start_matches("Uncaught ")
    .to_string()
}
```
```rust
for (description, js_error) in &summary.failures {
  // ...
  println!("{}", format_test_error(js_error));
  println!();
}
```

Comment (bartlomieju) [reply cli/fmt_errors.rs:265]: Makes sense, I'll do that in separate PR, so this change can be landed independently.

**Comment by lucacasonato** [comment]
Some feedback:

The `$name\n== output ==\n$output\n== output end ==\n$status ($time ms)\n` formatting seems weird to me. I think we should do something more like Rust does, where we capture and buffer by default, unless the test fails (in which case we dump the output with the error messages). A user could disable this behavior with a `--no-capture` flag or something (just like libtest in Rust).

The `running 3 tests from b_test.ts` messages are still very useless when running with `-j` flag. In parallel mode each test needs to be prefixed by test name, or test completions per file need to be buffered.

I don't like the many colors in the final output. I think there should only be colors for the FAILED / PASS label, not for the number of failures and passes.

---

Overall I think this takes some good inspiration from Vitest, but I think we should be more liberal with adopting some of their nice test output. I think for example that the extra left indentation and line numbers on the per test failure report is very nice:

image

---

I think we should also really consider more "radical" changes to the test runner output. For example I think including test names for each file by default is really not that important. If we instead had an "interactive mode" version of this output that showed the test file name, and then a live updating count of tests already run / left to run that'd be great IMO. That would also work out much better with the `-j` flag.

**Comment by nayeemrmn** [comment]
> image

The red background part in this image is a lot more readable than what we have now, especially on light theme it's not good without bold:
image 
image
or maybe 
image

File: cli/fmt_errors.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/fmt_errors.rs:265]: https://github.com/denoland/deno/pull/14302

**Comment by bartlomieju** [comment]
Updated output

**Comment by nayeemrmn** [comment]
I think the right-aligned text is out of place, this seem readable enough: 
image

**Comment by bartlomieju** [comment]
> I think the right-aligned text is out of place, this seem readable enough: image

Applied this suggestion

File: cli/tools/test.rs [hunk 2]
```diff
@@ -508,21 +502,101 @@ impl TestReporter for PrettyTestReporter {
...
+      println!(
+        "{}     {}{}",
+        colors::gray("ignored:"),
+        summary.ignored,
+        get_steps_text(summary.ignored_steps)
+      );
+    }
+    if summary.filtered_out > 0 {
+      println!("{}    {}", colors::gray("filtered:"), summary.filtered_out);
+    }
```
Comment (bartlomieju) [inline cli/tools/test.rs:532]: Keep in mind, that "failed", "ignored" and "filtered out" counts are conditional and are displayed only if the count of that category is greater than zero. Not sure if this is desirable or not. Options welcome

Comment (aapoalas) [reply cli/tools/test.rs:532]: Probably a good thing. There's very little reason to keep the cluttering messages around when they convey exactly no information.

**Pull request closed without merging**